### PR TITLE
Use member type for relational pattern even in error cases

### LINKED ISSUE #51936: Vs2019 v16.9.2 Raise runtime errors when using Property Pattern for undefined variable and fail compilation
**Version Used**: 
Microsoft Visual Studio Community 2019
Version 16.9.2

**Steps to Reproduce**:
 
1- Create a console net5 application.
2- Add the next class:

```
class Class1
    {
        public void Test1()
        {
           if (shape is Circle { Radius: >= 100 })  //variable shape is undefined and is using Property Relational Pattern, cause many errors by vs 2019.9.2
            {
                // this is a huge circle
            }
        }
    }
    public class Circle
    {
        public double Radius { get; set; }
    }
```

**Expected Behavior**:
In Editor, a red line under shape variable and a compilation error

**Actual Behavior**:
After step2 (without build), I get many repeated error messages at the upper screen under the menu with these common errors:
```
Feature 'Diagnostic analyzer runner' is currently unavailable due to an internal error. Show Stack Trace 
Feature 'CodeLens references' is currently unavailable due to an internal error. Show Stack Trace 
Feature 'Semantic classification cache' is currently unavailable due to an internal error. Show Stack Trace 

The common error message in Stack Trace:

RPC server exception:
System.InvalidOperationException: Operation is not valid due to the current state of the object.
```

as shown in the next figure:
vs2019-error

and I can't remove these errors.

When building I get the compilation error:

``` 
Error	MSB6006	"csc.exe" exited with code -2146232797.	Vs2019Bug	C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin\Roslyn\Microsoft.CSharp.Core.targets	71	

```

**More Details**

The line ( with undefined variable shape having a Property Check):
```
if (shape is Circle { Radius: >= 100 })
```
cause these errors, and it's expected that vs 2019 raise a compilation error.

If I Remove the Property Patten check and replace the above line with the next one:

```
if (shape is Circle ) //Vs works normally and cause compilation error as expected 
```

Vs 2019 is working fine and the build is done with a compilation error: 
```
error CS0103: The name 'shape' does not exist in the current context
1>Done building project "Vs2019Bug.csproj" -- FAILED.
```

  
**Update**
The Property Pattern above use  Property Relational patterns in C# 9.
If I Used Property patterns in C# 8 like:
```cs
  if (shape is Circle { Radius:  100 }) //property pattern of c#8, vs 2019 is working fine without above errors.
```
The vs 2019 is working fine:
```
error CS0103: The name 'shape' does not exist in the current context  
```

It seems that CodeAnalysis Crash with undefined variable using **Property Relational Pattern of C#9**

### PR DESCRIPTION
**Pull request by cston**
Fixes #51936

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs [hunk 1]
```diff
@@ -1193,10 +1192,12 @@ void addSubpatternsForTuple(ImmutableArray<TypeWithAnnotations> elementTypes)
...
 
-            if (inputType.IsErrorType() || hasErrors || symbol is null)
-                memberType = CreateErrorType();
-            else
-                memberType = symbol.GetTypeOrReturnType().Type;
+            memberType = symbol switch
+            {
+                FieldSymbol field => field.Type,
+                PropertySymbol property => property.Type,
+                _ => CreateErrorType()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:1199]: >CreateErrorType() [](start = 21, length = 17)

If this was causing a problem, is it still safe to use it for null result?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:1199]: >_ [](start = 16, length = 1)

It looks like there is a possible behavior change around other kinds of symbols. If other kinds not possible, consider explicitly checking for null and throwing for everything else.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 2)

File: src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs (hunk 1 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:1199]: > It looks like there is a possible behavior change around other kinds of symbols. 

There shouldn't be any change for other symbol kinds. `BindPropertyPatternMember()` sets `hasErrors` for symbols other than fields or properties and previously we used `CreateErrorType()` if `hasErrors` was set.

---
In reply to: [597137302](https://github.com/dotnet/roslyn/pull/51950#discussion_r597137302) [](ancestors = 597137302)

Comment (cston) [reply src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:1199]: > If this was causing a problem, is it still safe to use it for null result?

The problem was caused by an error type being used when the field or property type was available. In particular, the type returned from this method should match the type used for recursive property in `MakeTestsAndBindingsForRecursivePattern()`, which it should now.

---
In reply to: [597135186](https://github.com/dotnet/roslyn/pull/51950#discussion_r597135186) [](ancestors = 597135186)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 3)

**Comment by cston** [comment]
@dotnet/roslyn-compiler, please review, thanks.

**Comment by cston** [comment]
@dotnet/roslyn-compiler, please review, thanks.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 3)

**Pull request merged**